source/common/common/inline_map.h

```diff
@@ -411,44 +437,36 @@ template <class Key, class Value> class InlineMap : public InlineStorage {
     if (!inlineEntryValid(handle.inlineId())) {
       resetInlineMapEntry(handle.inlineId(), Value());
     }
-    return inline_entries_[handle.inlineId()];
-  }
-
-  /**
-   * Create an inline map with the given descriptor. If the descriptor is not finalized, it will be
-   * finalized before creating the inline map.
-   * @param descriptor the descriptor that contains the inline keys.
-   * @return the created inline map.
-   */
-  static std::unique_ptr<InlineMap> create(TypedInlineMapDescriptor& descriptor) {
-    if (!descriptor.finalized()) {
-      // Call finalize() to make sure that the descriptor is finalized and no any new inline
-      // keys could be registered.
-      descriptor.finalize();
-    }
-
-    return std::unique_ptr<InlineMap>(new ((descriptor.inlineKeysNum() * sizeof(Value)))
-                                          InlineMap(descriptor));
+    return inlineEntries()[handle.inlineId()];
   }
 
 private:
+  using StoragePtr = std::unique_ptr<uint8_t[]>;
   friend class InlineMapDescriptor<Key>;
 
-  InlineMap(TypedInlineMapDescriptor& descriptor) : descriptor_(descriptor) {
-    // Initialize the inline entries to nullptr.
-    uint64_t inline_keys_num = descriptor_.inlineKeysNum();
-    inline_entries_valid_.resize(inline_keys_num, false);
-    memset(inline_entries_storage_, 0, inline_keys_num * sizeof(Value));
-    inline_entries_ = reinterpret_cast<Value*>(inline_entries_storage_);
+  void lazyInitializeInlineEntriesMemory() {
+    if (inline_entries_ == nullptr) {
+      const uint64_t value_memory_size = descriptor_.inlineKeysNum() * sizeof(Value);
+      const uint64_t valid_memory_size = descriptor_.inlineKeysNum() * sizeof(bool);
+      const uint64_t memory_size = value_memory_size + valid_memory_size;
+
+      inline_entries_.reset(new uint8_t[memory_size]);
+      memset(inline_entries_.get(), 0, memory_size);
```

Comment:
Could you give me some more detail about the bug? I didn't get it accutally. I think this is more like the `reserve()` method. It reserver some memory but didn't use/initialize it.

Reply:
Oh I see; you don't have a bug-- you are using placement new below.

However you do not need to clear this memory. placement-new does not assume the memory is zeroed.

Reply:
The all bool flags need to be initialized to false, so this memset is still necessary.

Reply:
> suggest zeroing only the bool portion of this. Otherwise you could mask a failure to properly do the placement-new on the value-portion.

I forget this point. will update it.

> I'm still skeptical whether we need to have the bools and the values in the same allocation. Is there a perf benefit to that?

This at least reduce one time memory allocation. And I guess this should be a mature mode in the low-level container:

See the source code of absl raw_hash_set. (Although it's much complex than our inline map)
```
template <typename Alloc, size_t SizeOfSlot, size_t AlignOfSlot>
ABSL_ATTRIBUTE_NOINLINE void InitializeSlots(CommonFields& c, Alloc alloc) {
  assert(c.capacity_);
  // Folks with custom allocators often make unwarranted assumptions about the
  // behavior of their classes vis-a-vis trivial destructability and what
  // calls they will or won't make.  Avoid sampling for people with custom
  // allocators to get us out of this mess.  This is not a hard guarantee but
  // a workaround while we plan the exact guarantee we want to provide.
  const size_t sample_size =
      (std::is_same<Alloc, std::allocator<char>>::value && c.slots_ == nullptr)
          ? SizeOfSlot
          : 0;

  const size_t cap = c.capacity_;
  char* mem = static_cast<char*>(
      Allocate<AlignOfSlot>(&alloc, AllocSize(cap, SizeOfSlot, AlignOfSlot)));
  const GenerationType old_generation = c.generation();
  c.set_generation_ptr(
      reinterpret_cast<GenerationType*>(mem + GenerationOffset(cap)));
  c.set_generation(NextGeneration(old_generation));
  c.control_ = reinterpret_cast<ctrl_t*>(mem);
  c.slots_ = mem + SlotOffset(cap, AlignOfSlot);
  ResetCtrl(c, SizeOfSlot);
  if (sample_size) {
    c.infoz() = Sample(sample_size);
  }
  c.infoz().RecordStorageChanged(c.size_, cap);
}
```